File: ios/Classes/Serialization.swift
//
//  Serialization.swift
//  formx_sdk_flutter
//
//  Created by wu mark on 2023/12/19.
//

import Foundation
import FormX

extension FormXDataValue {
    func toMap() -> [String: Any?] {
        switch self {
        case .intValue(let value):
            return [
                "type": "FormXDataNumValue",
                "value": Double(value),
            ]
        case .doubleValue(let value):
            return [
                "type": "FormXDataNumValue",
                "value": value,
            ]
        case .listProduct(let productItems):
            return [
                "type": "FormXDataProductListValue",
                "value": productItems.map({ productItem in
                    [
                            "name": productItem.name ?? "",
                            "sku": productItem.sku ?? "",
                            "quantity": productItem.quantity ?? 0,
                            "amount": productItem.amount ?? 0.0,
                            "unitPrice": productItem.unitPrice ?? 0.0,
                            "discount": productItem.discount ?? "0",
                            ]
                })
            ]
        case .stringValue(let value):
            return [
                "type": "FormXDataStringValue",
                "value": value,
            ]
        case .intList(let list):
            return [
                "type": "FormXDataNumListValue",
                "value": list.map { Double($0) }
            ]
        case .doubleList(let list):
            return [
                "type": "FormXDataNumListValue",
                "value": list
            ]
        case .unsupported:
            return [
                "type": "FormXDataUnknownValue",
                "value": nil,
            ]
        case .boolValue(let value):
            return [
                "type": "FormXDataBoolValue",
                "value": value,
            ]
        case .boolList(let list):
            return [
                "type": "FormXDataBoolListValue",
                "value": list,
            ]
        case .stringList(let list):
            return [
                "type": "FormXDataStringListValue",
                "value": list,
            ]
        case .map(let map):
            return [
                "type": "FormXDataMapValue",
                "value": map,
            ]
        case .mapList(let list):
            return [
                "type": "FormXDataMapListValue",
                "value": list,
            ]
        }
    }
}

extension ExtractMetaData {
    func toMap() -> [String: Any?] {
        return [
            "extractorId": extractorId,
            "jobId": jobId,
            "usage": usage,
            "requestId": requestId,
        ]
    }
}

extension FormXData {
    func toMap() -> [String: Any?] {
      return [
        "value": value.toMap(),
        "valueType": valueType,
        "confidence": confidence,
        "extractedBy": extractedBy
      ]
    }
}

extension ExtractDocumentMetaData {
    
    func toMap() -> [String: Any?] {
        return [
            "extractorType": extractorType,
            "pageNo": pageNo,
            "sliceNo": sliceNo,
            "orientation": orientation
        ]
    }
}

extension ExtractDocument {
    func toMap() -> [String: Any?] {
      return [
        "type": "ExtractDocument",
        "extractorId": extractorId,
        "documentType": type,
        "typeConfidence": typeConfidence,
        "metaData": metadata.toMap(),
        "data": data.toMap(),
        "detailedData": detailedData.mapValues({ d in
            d.map { $0.toMap() }
        }),
        "boundingBox": boundingBox
      ]
    }
}

extension ExtractDocumentError.Error {
    func toMap() -> [String: Any?] {
        return [
            "code": code,
            "message": message,
            "info": info,
        ]
    }
}

extension ExtractDocumentError {
    func toMap() -> [String: Any?] {
        return [
            "type": "ExtractDocumentError",
            "extractorId": extractorId,
            "metaData": metaData.toMap(),
            "error": error.toMap()
        ]
    }
}

extension FormXAPIExtractResult {
    func toMap() -> [String: Any] {
        return [
            "status": self.response.status,
            "metaData": self.response.metadata.toMap(),
            "documents": self.response.documents.map({ result in
                switch result {
                case .failed(let error):
                    error.toMap()
                case .success(let doc):
                    doc.toMap()
                }
            })
        ]
    }
}

extension FormXAPIDetectDocumentsResult {
    func toMap() -> [String: Any] {
        return [
            "status": response.status,
            "documents": response.documents.map({ region in
                return [
                    "bbox": [
                        "left": region.bbox[0],
                        "top": region.bbox[1],
                        "right": region.bbox[2],
                        "bottom": region.bbox[3],
                    ],
                    "type": region.type,
                ]
            }),
        ]
    }
}

extension FormXError {
    
    func toMap() -> [String: Any] {
      return [
        "code": self.code,
        "message": self.message,
      ]
    }
}

extension [String:FormXData?] {
    func toMap() -> [String: Any?] {
        return self.mapValues { v in
            if let v = v {
                return [
                    "confidence": v.confidence,
                    "value": v.value.toMap(),
                    "valueType": v.valueType,
                    "extractedBy": v.extractedBy
                ]
            }
            return nil
        }
    }
}

File: android/src/main/kotlin/ai/formx/mobile/sdk/formx_sdk_flutter/Serialization.kt
package ai.formx.mobile.sdk.formx_sdk_flutter

import ai.formx.mobile.sdk.ExtractDocument
import ai.formx.mobile.sdk.ExtractDocumentError
import ai.formx.mobile.sdk.ExtractDocumentMetaData
import ai.formx.mobile.sdk.ExtractDocumentResult
import ai.formx.mobile.sdk.ExtractMetaData
import ai.formx.mobile.sdk.FormXAPIDetectDocumentsResponse
import ai.formx.mobile.sdk.FormXAPIExtractResponse
import ai.formx.mobile.sdk.FormXDataBoolListValue
import ai.formx.mobile.sdk.FormXDataBoolValue
import ai.formx.mobile.sdk.FormXDataMapListValue
import ai.formx.mobile.sdk.FormXDataMapValue
import ai.formx.mobile.sdk.FormXDataNumListValue
import ai.formx.mobile.sdk.FormXDataNumValue
import ai.formx.mobile.sdk.FormXDataProductListValue
import ai.formx.mobile.sdk.FormXDataStringListValue
import ai.formx.mobile.sdk.FormXDataStringValue
import ai.formx.mobile.sdk.FormXDataUnknownValue
import ai.formx.mobile.sdk.FormXDataValue
import ai.formx.mobile.sdk.FormXProductValue

fun FormXAPIDetectDocumentsResponse.toMap(): HashMap<String, Any> =
    HashMap<String, Any>().apply {
        put("status", status)
        put("documents",
            documents.map {
                HashMap<String, Any>().apply {
                    put("type", it.type)
                    put("bbox", HashMap<String, Int>().apply {
                        put("left", it.bbox[0])
                        put("top", it.bbox[1])
                        put("right", it.bbox[2])
                        put("bottom", it.bbox[3])
                    })
                }
            }
        )
    }

fun FormXAPIExtractResponse.toMap(): HashMap<String, Any> =
    HashMap<String, Any>().apply {
        put("status", status)
        put("metaData", metaData.toMap())
        put("documents", documents.map {
            it.toMap()
        })
    }

fun ExtractMetaData.toMap(): HashMap<String, Any> =
    HashMap<String, Any>().apply {
        put("extractorId", extractorId)
        put("requestId", requestId)
        put("usage", usage)
        jobId?.let {
            put("jobId", it)
        }
    }

fun ExtractDocument.toMap(): HashMap<String, Any> =
    HashMap<String, Any>().apply {
        put("type", "ExtractDocument")
        put("extractorId", extractorId)
        put("metaData", metaData.toMap())
        type?.let {

            put("documentType", it)
        }
        typeConfidence?.let {
            put("typeConfidence", it)
        }
        put("data", data.mapValues {
            it.value?.toMap()
        })
        put("detailedData", detailedData.mapValues {
            it.value.map { v ->
                v.toMap()
            }
        })
        boundingBox?.let {
            put("boundingBox", it)
        }
    }

fun ExtractDocumentMetaData.toMap(): HashMap<String, Any> =
    HashMap<String, Any>().apply {
        extractorType?.let {
            put("extractorType", it)
        }
        put("pageNo", pageNo)
        put("sliceNo", sliceNo)
    }

fun FormXProductValue.toMap(): HashMap<String, Any?> =
    HashMap<String, Any?>().apply {
        put("name", name)
        put("amount", amount ?: 0.0)
        put("discount", discount)
        put("sku", sku)
        put("quantity", quantity ?: 0)
        put("unitPrice", unitPrice ?: 0.0)
    }

fun FormXDataValue.toMap(): HashMap<String, Any?> =
    HashMap<String, Any?>().also { it ->
        it["extractedBy"] = extractedBy
        it["confidence"] = confidence
        it["valueType"] = valueType
        it["value"] = HashMap<String, Any?>().also { valueMap ->
            valueMap["type"] = this.javaClass.simpleName
            when (this) {
                is FormXDataBoolListValue -> valueMap["value"] = value
                is FormXDataBoolValue -> valueMap["value"] = value
                is FormXDataMapListValue -> valueMap["value"] = value
                is FormXDataMapValue -> valueMap["value"] = value
                is FormXDataNumListValue -> valueMap["value"] = value.map { it.toDouble() }
                is FormXDataNumValue -> {
                  valueMap["value"] = value.toDouble()
                }
                is FormXDataProductListValue -> valueMap["value"] = value.map { productValue ->
                    productValue.toMap()
                }

                is FormXDataStringListValue -> valueMap["value"] = value
                is FormXDataStringValue -> valueMap["value"] = value
                is FormXDataUnknownValue -> valueMap["value"] = value
            }
        }
    }

fun Error.toMap(): HashMap<String, Any> =
    HashMap<String, Any>().apply {
        put("code", code.name)
        put("message", message ?: "")
    }

fun ai.formx.mobile.sdk.Error.toMap(): HashMap<String, Any> =
    HashMap<String, Any>().apply {
        put("code", code)
        put("message", message)
        info?.let {
            put("info", it)
        }
    }

fun ExtractDocumentError.toMap(): HashMap<String, Any> =
    HashMap<String, Any>().apply {
        put("type", "ExtractDocumentError")
        put("metaData", metaData.toMap())
        extractorId?.let {
            put("extractorId", it)
        }
        put("error", error.toMap())
    }

fun ExtractDocumentResult.toMap(): HashMap<String, Any> =
    when (this) {
        is ExtractDocument -> this.toMap()
        is ExtractDocumentError -> this.toMap()
    }
